neo/Consensus/ConsensusService.cs

```diff
@@ -169,37 +169,32 @@ private void Log(string message, LogLevel level = LogLevel.Info)
+            // and issues a change view for the same view, it will have a different hash and will correctly respond
+            // again; however replay attacks of the ChangeView message from arbitrary nodes will not trigger an
+            // additonal recovery message response.
+            if (!knownHashes.Add(payload.Hash)) return;
+            if (message.NewViewNumber <= context.ViewNumber)
             {
-                if (!shouldSendRecovery)
+                bool shouldSendRecovery = false;
+                // Limit recovery to sending from `f` nodes when the request is from a lower view number.
+                int allowedRecoveryNodeCount = context.F;
+                for (int i = 0; i < allowedRecoveryNodeCount; i++)
                 {
-                    // Limit recovery to sending from `f` nodes when the request is from a lower view number.
-                    int allowedRecoveryNodeCount = context.F;
-                    for (int i = 0; i < allowedRecoveryNodeCount; i++)
-                    {
-                        var eligibleResponders = context.Validators.Length - 1;
-                        var chosenIndex = (payload.ValidatorIndex + i + message.NewViewNumber) % eligibleResponders;
-                        if (chosenIndex >= payload.ValidatorIndex) chosenIndex++;
-                        if (chosenIndex != context.MyIndex) continue;
-                        shouldSendRecovery = true;
-                        break;
-                    }
+                    var eligibleResponders = context.Validators.Length - 1;
+                    var chosenIndex = (payload.ValidatorIndex + i + message.NewViewNumber) % eligibleResponders;
+                    if (chosenIndex >= payload.ValidatorIndex) chosenIndex++;
+                    if (chosenIndex != context.MyIndex) continue;
+                    shouldSendRecovery = true;
+                    break;
                 }
 
-                // We keep track of the payload hashes received in this block, and don't respond with recovery
-                // in response to the same payload that we already responded to previously.
-                // ChangeView messages include a Timestamp when the change view is sent, thus if a node restarts
-                // and issues a change view for the same view, it will have a different hash and will correctly respond
-                // again; however replay attacks of the ChangeView message from arbitrary nodes will not trigger an
-                // additonal recovery message response.
-                if (!shouldSendRecovery || knownHashes.Contains(payload.Hash)) return;
-                knownHashes.Add(payload.Hash);
+                if (!shouldSendRecovery) return;
 
                 Log($"send recovery from view: {message.ViewNumber} to view: {context.ViewNumber}");
                 localNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeRecoveryMessage() });
-                return;
```

Comment:
It seems like it doesn’t hurt to return here if we sent recovery; won’t the next if just return anyway?

Reply:
https://github.com/neo-project/neo/blob/3aa09896b591a44f0e45a3c5cfaa92778d0896ee/neo/Consensus/ConsensusService.cs#L200-L202

The `expectedView` may less then `context.ViewNumber`.

Reply:
Yes that was my point, but if you return in the branch it avoids that one check again before returning. It doesn’t matter much. I’m resolving the conversation.